Expose token_signing_key option

aeijdenberg opened the pull request:
This would be helpful for [issue 1559](https://github.com/concourse/concourse/issues/1559) as it would allow us to generate our own token signing key and pass it through the appropriate jobs using the existing option flags.

jtarchie commented:
@aeijdenberg, were there more changes with this? This looks like this are only bosh property changes.

aeijdenberg commented:
@jtarchie - no, this really is it. The option to accept a signing key already exists in `atc` and `tsa` - this patch just allows it to be set via a bosh manifest.

I do have [this related pull request](https://github.com/concourse/fly/pull/193) to `fly`, and together that gives us the primitives we need for a reasonable SSO solution.

vito commented on jobs/atc/spec:
```diff
@@ -478,3 +479,11 @@ properties:
+    description: |
+      PEM RSA private key used for minting ATC tokens. If not set, one is generated
+      automatically.
+    default:
```
I think it'd be better to omit `default` since the default isn't actually respected. You could then use `if_p` to check if it's configured.

Example: https://github.com/concourse/concourse/blob/9b041c529a54242f2ce69fae6d31ba9c0ebce31d/jobs/atc/templates/atc_ctl.erb#L96

vito commented:
@aeijdenberg Do you have time to update this? (Thanks!)

aeijdenberg commented:
Hi @jtarchie - yes, actually I got back onto this yesterday. Sadly I lack knowledge of Ruby and ERBs (and don't have as efficient as I'd like debug/test/release cycle) so am a bit slow. Will try to push a new commit today, and will rebase on master.

One question though - have we considered simply removing the `generated_signing_key` package and instead making this a required field?

It would be trivial to have it [auto-generated](https://bosh.io/docs/variable-types.html#rsa) by a bosh manifest, e.g. 

```yaml
instance_groups:
- name: web
  jobs:
  - name: atc
    properties:
      token_signing_key: ((token_signing_key))
  - name: tsa
    properties:
      token_signing_key: ((token_signing_key))

...

variables:
- name: token_signing_key
  type: rsa
```

vito commented:
@aeijdenberg It's worth considering. We could probably do the same with the `generated_worker_key` and `generated_tsa_host_key` packages. You often need to change them before you do anything more serious anyway.

The original goal was to simplify things for people getting started, but yeah, generated variables can accomplish that now. Plus CredHub. And people using the binaries have to specify them anyway.

aeijdenberg commented:
I suspect the current config also has the side-effect that if multiple Concourses were deployed on the same bosh, that all might end up sharing the same keys - since that's effectively built at compile time (and it's perhaps surprising that credential is generated at compile time - one would normally not consider those artefacts to contain secrets?).

Why don't I "pivot" this PR to make them all configurable via a manifest?

aeijdenberg commented:
@vito - have pushed new commit that removes all 3 `generated_xxx` packages, and replaces their use with bosh variables.

Updated the `single-vm.yml` manifest to specify these, and also updated it to work a bit more cleanly on a more "standard" bosh-lite or other cloud-config.

Seems to deploy cleanly in our real env too.

vito commented:
Sorry for the wait, gonna take a look at this now!

vito commented:
I get this error when trying to run `./dev/vbox/deploy` with these changes:

```
17:02:54 | Preparing deployment: Preparing deployment (00:00:00)
            L Error: Failed to generate variable '/Bosh Lite Director/concourse/postgres_password' from config server: Director is not configured with a config server

17:02:54 | Error: Failed to generate variable '/Bosh Lite Director/concourse/postgres_password' from config server: Director is not configured with a config server

Started  Tue Nov 21 17:02:54 UTC 2017
Finished Tue Nov 21 17:02:54 UTC 2017
Duration 00:00:00

Task 5 error

Updating deployment:
  Expected task '5' to succeed but state is 'error'

Exit code 1
```

Looks like it doesn't know to generate them client-side. Bummer.

vito commented:
Ah! Got past that by adding `--vars-store` to the `deploy` script. Makes sense.

vito commented on jobs/groundcrew/spec:
```diff
@@ -157,3 +158,26 @@ properties:
+       public_key: |
+          ssh-rsa ...
+
+  worker_key:
```
This shouldn't be necessary; we already have `tsa.private_key`. I can just omit it when merging.

vito commented on jobs/groundcrew/spec:
```diff
@@ -157,3 +158,26 @@ properties:
+
+  tsa_host_key:
+     description: |
+       Must be specified, bosh can auto-generate (type ssh), see sample manifest.yml.
```
I think you can specify types in the job spec now. I can do that too.

vito commented on jobs/groundcrew/templates/drain.erb:
```diff
@@ -33,7 +36,7 @@ fi
 %>
 
-for key in /var/vcap/packages/generated_worker_key/id_rsa \
+for key in $WORKER_KEY \
```
Don't need a loop here anymore.

vito commented on jobs/tsa/spec:
```diff
@@ -98,3 +98,40 @@ properties:
+       public_key: |
+          ssh-rsa ...
+
+  worker_key:
```
Hmm, this overlaps with `authorized_keys`. I think we can just remove it.

vito commented on manifests/single-vm.yml:
```diff
@@ -1,11 +1,11 @@
 ---
-name: ((deployment_name))
+name: concourse
```
Don't want this change.

vito commented on manifests/single-vm.yml:
```diff
@@ -1,11 +1,11 @@
 
 releases:
 - name: concourse
-  version: ((concourse_version))
```
Or these.

vito commented:
thanks! i've pulled in your changes and updated them. i made #1834 to track the remaining work.

aeijdenberg commented:
Thanks @vito - glad it was helpful!

bmhkb4 commented:
I'm also getting :  Error: Failed to generate variable '/standalone_bosh/concourse/token_signing_key' from config server: Director is not configured with a config server

What do you mean by 

> "Got past that by adding --vars-store to the deploy script. Makes sense." 

Can you perhaps also update the [release notes](http://concourse-ci.org/downloads.html#v370) as I have spent many hours trying to figure this out until I stumbled on this thread.  It isn't clear in the release notes what to do to make this work and the referenced [github link](http://github.com/concourse/concourse/issues/1834) doesn't actually explain what to do.

bmhkb4 commented:
Figured it out, for those of you searching for the answer.  Using [this](https://bosh.io/docs/cli-int.html) and [this](https://bosh.io/docs/variable-types.html) and [this](https://gist.github.com/vito/44cdf3aafcf4ba06349e24e8c21135eb) I was able to piece together a fix.

The answer is two-fold.  The first thing is update your concourse.yml to include [these](https://gist.github.com/vito/44cdf3aafcf4ba06349e24e8c21135eb), literally just add them to your manifest as they are.  The second, more obtuse part, is to provide a dummy "--vars-store=creds.yml" attribute in your concourse deployment command.  As far as I can tell, it doesn't actually need to exist, you just need to set the var store variable to something (so it passes syntax checks on the deploy).  Then, run your re-deploy and it will programmatically generate the variables in your concourse.yml and voila,  a working upgrade to >3.7 concourse.

Example:
`/usr/local/bin/bosh -e ${DIRECTOR_NAME} deploy -d concourse --no-redact ${TMP} \
  --vars-store=creds.yml \
  -v concourse-director-uuid=${BOSH_DIRECTOR_UUID} \
  -v concourse-singleton-az=${CONCOURSE_SINGLETON_AZ} \
  -v concourse-worker-azs=${CONCOURSE_WORKER_AZS} \
  -v concourse-network-name=${CONCOURSE_NETWORK_NAME} \
  -v concourse-web-static-ips=${CONCOURSE_WEB_STATIC_IPS} \
  -v concourse-web-url=${CONCOURSE_WEB_URL} \
  -v concourse-web-basic-auth-username=${CONCOURSE_WEB_BASIC_AUTH_USERNAME} \
  -v concourse-web-vault-url=${CONCOURSE_VAULT_URL} \`
